airflow/utils/setup_teardown.py

```diff
@@ -134,57 +168,147 @@ def push_setup_teardown_task(cls, operator: AbstractOperator | list[AbstractOper
+        else:
+            cls.set_dependency(roots, setup_task, upstream=False)
 
     @classmethod
     def set_work_task_roots_and_leaves(cls):
-        if setup_task := cls.get_context_managed_setup_task():
+        """Sets the work task roots and leaves."""
+        if setup_task := cls._context_managed_setup_task:
             if isinstance(setup_task, list):
                 setup_task = tuple(setup_task)
-            tasks_in_context = cls.context_map.get(setup_task, [])
+            tasks_in_context = [
+                x for x in cls.context_map.get(setup_task, []) if not x.is_teardown and not x.is_setup
+            ]
             if tasks_in_context:
                 roots = [task for task in tasks_in_context if not task.upstream_list]
                 if not roots:
-                    setup_task >> tasks_in_context[0]
-                elif isinstance(setup_task, tuple):
-                    for task in setup_task:
-                        task >> roots
+                    setup_task >> list(tasks_in_context)[0]
                 else:
-                    setup_task >> roots
-        if teardown_task := cls.get_context_managed_teardown_task():
+                    cls.set_dependency(roots, setup_task, upstream=False)
+                leaves = [task for task in tasks_in_context if not task.downstream_list]
+                cls.set_teardown_task_as_leaves(leaves)
+
+        if teardown_task := cls._context_managed_teardown_task:
             if isinstance(teardown_task, list):
                 teardown_task = tuple(teardown_task)
-            tasks_in_context = cls.context_map.get(teardown_task, [])
+            tasks_in_context = [
+                x for x in cls.context_map.get(teardown_task, []) if not x.is_teardown and not x.is_setup
+            ]
             if tasks_in_context:
                 leaves = [task for task in tasks_in_context if not task.downstream_list]
                 if not leaves:
-                    teardown_task << tasks_in_context[-1]
-                elif isinstance(teardown_task, tuple):
-                    for task in teardown_task:
-                        task << leaves
+                    teardown_task << list(tasks_in_context).pop()
                 else:
-                    teardown_task << leaves
+                    cls.set_dependency(leaves, teardown_task)
+                roots = [task for task in tasks_in_context if not task.upstream_list]
+                cls.set_setup_task_as_roots(roots)
+        cls.set_setup_teardown_relationships()
+        cls.active = False
+
+    @classmethod
+    def set_setup_teardown_relationships(cls):
+        """
+        Here we set relationship between setup to setup and
+        teardown to teardown.
+
+        code:: python
+            with setuptask >> teardowntask:
+                with setuptask2 >> teardowntask2:
+                    ...
+
+        We set setuptask >> setuptask2, teardowntask >> teardowntask2
```

Comment:
it may be out of the scope for this PR, but i think the behavior should be that we don't arrow teardowntask to teardowntask2 -- by default i think they should run in parallel.  wdyt?  making this change would make it consistent with how we handle teardowns in task groups -- they are ignored when calculating leaves.